tests/core/escaping/main.fmf

```diff
@@ -6,11 +6,7 @@ description:
 test: ./test.sh
 framework: beakerlib
 tag-: [container, virtual]
+tag+: [provision-virtual]
 
 environment:
     PROVISION_METHODS: local container
-adjust:
```

Comment:
I don't understand why this changed.

Reply:
My idea was to simplify the setup and get rid of the old full test suite. Perhaps, it's too early? But in the end, I believe, these should not be needed. Each test would just keep their default list of provision methods, quite often probably just `local` and the other provision methods would be excercised through the dedicated provision plans which will select suitable candidates by the `provision-*` tag.

Reply:
I see. I guess the puzzling part was mostly "why this test and no other". Makes sense.

How about tests & features that make no sense for every provisioning method? E.g. something related to SSH won't apply to `local` or `container`. A `continue` in the `for method in $PROVISION_METHODS` loop in those tests, or is it possible to drop such methods on the test level?

Reply:
> I see. I guess the puzzling part was mostly "why this test and no other". Makes sense.

Yeah, just experimenting to confirm it actually works. Converted to draft as it's more appropriate.
 
> How about tests & features that make no sense for every provisioning method? E.g. something related to SSH won't apply to `local` or `container`. A `continue` in the `for method in $PROVISION_METHODS` loop in those tests, or is it possible to drop such methods on the test level?

Tests would apply only those `provision-*` tags which are relevant/supported. So they wouldn't / shouldn't be called with unsupported methods. Plus, yes, I think each test can skip certain part of the code using `continue` if irrelevant for given provision method. If a provision method is not supported at all I'd suggest the test should error.

Reply:
Could we have more tags and also a negative? We already know that ssh/container/local differ for several features - so If test doesn't make sense on eg. local it could set `tag: no-local`, similar to test which needs ssh so `tag: ssh`. 

And thus the provision methods plan can do e.g.
- connect: tag:ssh,connect -tag:no-ssh -tag:no-connect

Without negative tests we would need to update test whenever new provision method is added (e.g. our internal minute is  `tag:ssh`).

Reply:
Hm, I was heading into direction of simplifying the setup as the current situation seems quite confusing to me. So the idea was that `provision` plugin would just filter on a single tag:

https://github.com/teemtee/tmt/blob/068ec64f6cd6b5d17763e43bc786e7f88ea61602/plans/provision/virtual.fmf#L12-L14

And for this setup I don't see how negative tags would/could be used. And yes, I would probably rather welcome explicitly adding any newly added provision method to all relevant tests which should be exercises by it.

Also inheritance could be used when/if possible. If a particular test from a larger test directory does not fit given provision method it would just remove the tag.

> * connect: tag:ssh,connect -tag:no-ssh -tag:no-connect

I'm afraid I don't understand this use case. And it seems to be quite complex to read. Could you add some more explanation or an example?

Reply:
(I've been asked for an opinion, so, doing so...)

I for one am not that invested in the outcome of this particular thread, I can imagine both ways to work successfully. I considered @psss's patch as a good starting point we could then build on. I tried to untangle the `PROVISION_METHOD` use in our tests, and I failed, mostly because of how everything is connected to `how=full` and `trigger=commit` checks, which I find to be a major blocker to any simplification. So, any simplification would be nice, and this one seemed to be in the right direction.

The necessity of tagging all tests when a new plugin arrives is a valid point, although, as long as we can list all tests parametrized by `PROVISION_METHODS`, it would be just tedious rather than impossible. Not great, not terrible. But I still think this would be possible to remedy once we untangle tests parametrized by provisioning, and make it easier to digest.

Reply:
The idea was to tag tests by 'category' of provision methods.
Use case:  Testing bug happening only when ssh is used - such test should run on 'virtual, connect, minute, beaker' but can't (or doesn't make sense to) run on 'local, container'.

If I use current proposal I would tag it by `tag: [ provision-virtual, provision-connect, provision-minute, provision-beaker]` (and btw, provision-virtual is already a short name for provision-virtual-testcloud. Which might be split into two once provision-virtual-libvirt is available).

The category tags can be used as 'run on all provision methods of "ssh" category, test doesn't care which'
I agree that the negative tags might be overkill right now.

Will we tag tests which are parametrized by `PROVISION_METHODS` ? I guess the default provision in tests is 'local'.

Reply:
Discussed on the hacking session and today, new tags described in 249bef178c87b67d152b2e13d2eb6b804c4df4b0.